src/frontend/js/project-service.js

```diff
@@ -577,11 +712,48 @@ angular.module('seashell-projects', ['seashell-websocket', 'marmoset-bindings'])
           var self = this;
           // TODO: handle racket files.
           var tests = test ? self.getTestsForQuestion(question) : [];
+          var path = self._getPath(question, "question", self.fileToRun);
+          var file = self.root.find(path);
 
+          if(!file) {
+            return $q.reject("Attempting to run file that does not exist.");
+          }
           if (test && tests.length === 0)
             return $q.reject("No tests for question!");
 
-          return $q.when(ws.compileAndRunProject(self.name, question, tests));
+          if(ws.connected && settings.settings.offline_mode !== 2) {
+            return $q.when(ws.compileAndRunProject(self.name, question, tests));
+          }
+          else if(settings.settings.offline_mode === 2 || !ws.connected && settings.settings.offline_mode !== 0) {
+            var res = $q.defer();
+            if(!self.compiler) {
+              self.compiler = new Worker("js/offline-compile.js");
+            }
+            self.compiler.onmessage = function(result) {
+              if(result.data.status == "compile-failed") {
+                res.reject(result.data);
+              }
+              else if(result.data.status == "running") {
+                self.runner = new Worker("js/offline-run.js");
```

Comment:
Should have a service to manage the list of running "programs" (runner Worker instances) and "PIDs" so we can actually implement kill.